File: src/main/scala/com/dr_benway/tknowledge/lib/crafting/FountainRecipe.scala
package com.dr_benway.tknowledge.lib.crafting

import net.minecraft.item.ItemStack
import thaumcraft.api.aspects.AspectList
import net.minecraftforge.fluids.FluidStack
import net.minecraftforge.oredict.OreDictionary
import thaumcraft.api.aspects.Aspect
import net.minecraftforge.fluids.Fluid
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.nbt.NBTTagString
import net.minecraft.entity.player.EntityPlayer
import thaumcraft.api.research.ResearchHelper



class FountainRecipe(research1: String, key1: String, output: FluidStack, input: Any, itags: AspectList, aura: AspectList, danger: Float) {
  
  
  val cat = input match {
    case s: String => OreDictionary.getOres(input.asInstanceOf[String])
    case null => null
    case _ => input.asInstanceOf[ItemStack]
  }
  
  val research = research1
  val key = key1
  val aspects = itags
  val result = output
  val vis_aspects = aura
  val pollutingChance = if(danger >= 0.0F && danger <= 1.0F) danger else if(danger > 1.0F) 1.0F else 0.0F //yeah there is a smarter way
  
  def matches(in: ItemStack, out: Fluid, player: EntityPlayer): Boolean = {
    if(out == null) false
    if( (cat.isInstanceOf[ItemStack]) && (OreDictionary.itemMatches(cat.asInstanceOf[ItemStack], in, false)) ||
      ((cat.isInstanceOf[Seq[_]]) && (cat.asInstanceOf[Seq[_]].size > 0)) ) {
      result.isInstanceOf[FluidStack] && result.getFluid == out && ResearchHelper.isResearchComplete(player.getName, this.research)
    } else false
  }
  
  
  
}

File: src/main/scala/com/dr_benway/tknowledge/api/FountainRecipe.java
package com.dr_benway.tknowledge.api;

import net.minecraft.item.ItemStack;
import thaumcraft.api.aspects.AspectList;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.oredict.OreDictionary;
import net.minecraftforge.fluids.Fluid;
import net.minecraft.entity.player.EntityPlayer;
import thaumcraft.api.research.ResearchHelper;

public class FountainRecipe {
	
	private FluidStack result;
	private Object cat;
	private AspectList aspects;
	private AspectList vis_aspects;
	private String research;
	private String key;
	private float pollutingChance;
	
	
	public FountainRecipe(String research1, String key1, FluidStack output, Object input, AspectList itags, AspectList aura, float danger) {
		
		this.cat = input;
		if(input instanceof String)
			this.cat = OreDictionary.getOres((String)input);
		this.research = research1;
		this.key = key1;
		this.result = output;
		this.aspects = itags;
		this.vis_aspects = aura;
		if(danger >= 0.0 && danger <= 1.0) {
			this.pollutingChance = danger;
		} else if(danger >= 1.0) {
				this.pollutingChance = 1.0F;
			} else this.pollutingChance = 0.0F;
				
		
	}
	
	
	public boolean matches(ItemStack in, Fluid out, EntityPlayer player) {
		if(out == null || in == null) return false;
		if(this.cat instanceof ItemStack && OreDictionary.itemMatches((ItemStack)cat, in, false)) {	
			return (this.result.getFluid() == out && ResearchHelper.isResearchComplete(player.getName(), this.research));
		} else { return false; }
	}
	
	
	
	public FluidStack getResult() {
		return this.result;
	}
	
	public Object getCatalyst() {
		return this.cat;
	}
	
	public String getResearch() {
		return this.research;
	}
	
	public String getKey() {
		return this.key;
	}
	
	public AspectList getEssentia() {
		return this.aspects;
	}
	
	public AspectList getVis() {
		return this.vis_aspects;
	}
	
	public float getPollutingChance() {
		return this.pollutingChance;
	}

}
